### DISCUSSION TITLE
Initial tokens after user import

### DISCUSSION THREAD

**Discussion by quickstar** [ory/kratos#2696] [Q&A]
Hi everyone

We're building a system which is invitation based, pretty similar to the one explained in this ticket https://github.com/ory/kratos/issues/2525.

In our case an administrator connects the application to an azure tenant to fetch a list with all users in the azure AD. The admin can then select users which he wants to invite. For every invited user we create a kratos identity. Thanks to https://github.com/ory/kratos/pull/2347 we are now able to import the Social sign-in connection directly. This works just fine and the user is able to directly login to kratos.

Unfortunately, since the user didn't link his azure account himself, kratos doesn't have the initial tokens (initial_id_token, initial_access_token, initial_refresh_token).

We need the token material since we provide a user delegated access to further ms graph resources. 

What would be a clean way to get those tokens?

Since kratos doesn't refresh the tokens, we've built a microservice to just do that. But if for example this service crashes and our tokens expire we don't have a clean way of obtaining new token material.

Therefore would it make sense to introduce a flow which allows kratos to fetch new token material?

**Comment by aeneasr** [comment]
That would probably be solved by: https://github.com/ory/kratos/pull/2428

**Reply by quickstar** [reply]
This looks very promising! If I understand the code correctly this would also solve the first case

> Unfortunately, since the user didn't link his azure account himself, kratos doesn't have the initial tokens (initial_id_token, initial_access_token, initial_refresh_token).

right?

**Reply by david972** [reply]
hi yes this is the goal of my pull-request
